### PR TITLE
Reapply " [XRay] Add support for instrumentation of DSOs on x86_64 (#90959)"

### PR DESCRIPTION
**Pull request by sebastiankreutzer**
This fixes remaining issues in my previous PR #90959.

Changes:
- Removed dependency on LLVM header in `xray_interface.cpp`
- Fixed XRay patching for some targets due to missing changes in architecture-specific patching functions
- Addressed some remaining compiler warnings that I missed in the previous patch
- Formatting

I have tested these changes on `x86_64` (natively), as well as `ppc64le`, `aarch64` and `arm32` (cross-compiled and emulated using qemu).

**Original description:**

This PR introduces shared library (DSO) support for XRay based on a revised version of the implementation outlined in [this RFC](https://discourse.llvm.org/t/rfc-upstreaming-dso-instrumentation-support-for-xray/73000).
The feature enables the patching and handling of events from DSOs, supporting both libraries linked at startup or explicitly loaded, e.g. via `dlopen`.
This patch adds the following:
- The `-fxray-shared` flag to enable the feature (turned off by default)
- A small runtime library that is linked into every instrumented DSO, providing position-independent trampolines and code to register with the main XRay runtime
- Changes to the XRay runtime to support management and patching of multiple objects

These changes are fully backward compatible, i.e. running without instrumented DSOs will produce identical traces (in terms of recorded function IDs) to the previous implementation.

Due to my limited ability to test on other architectures, this feature is only implemented and tested with x86_64. Extending support to other architectures is fairly straightforward, requiring only a position-independent implementation of the architecture-specific trampoline implementation (see `compiler-rt/lib/xray/xray_trampoline_x86_64.S` for reference). 

This patch does not include any functionality to resolve function IDs from DSOs for the provided logging/tracing modes. These modes still work and will record calls from DSOs, but symbol resolution for these functions in not available. Getting this to work properly requires recording information about the loaded DSOs and should IMO be discussed in a separate RFC, as there are mulitple feasible approaches.

### CODE REVIEW

**Comment by sebastiankreutzer** [comment]
Buildkite failure is due to unrelated errors in `libcxx/test/libcxx/headers_in_modulemap.sh.py`

**Review by jplehr — APPROVED** [review]
LGTM

File: clang/test/Driver/XRay/xray-shared.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// RUN: %clang -### --target=x86_64-unknown-linux-gnu -fPIC -fxray-instrument -fxray-shared -c %s -o /dev/null 2>&1 | FileCheck %s
+// RUN: %clang -### --target=x86_64-unknown-linux-gnu -fpic -fxray-instrument -fxray-shared -c %s -o /dev/null 2>&1 | FileCheck %s
+// RUN: %clang -### --target=x86_64-unknown-linux-gnu -fxray-instrument -fxray-shared -c %s -o /dev/null 2>&1 | FileCheck %s
```
Comment (rupprecht) [inline clang/test/Driver/XRay/xray-shared.cpp:3]: This test case fails for builds where the `CLANG_DEFAULT_PIE_ON_LINUX` cmake option is not the default.

If this test line is important, we could move it to a separate test file guarded by the `default-pie-on-linux` lit feature, i.e. like https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/linux-default-pie.c.

However, I'd argue this case is not worth keeping. The two test lines above verify `-fxray-shared` work w/ pic, and the two below verify it doesn't work w/ no-pic. With the tests surrounding this in place, the only thing this verifies is that pic is the default (which isn't always true), and linux-default-pie.c already covers that behavior.

Any objection to just removing this line? Or should I move it to a separate file? (btw, not asking for a revert -- I'm just ignoring this test failure)

Comment (sebastiankreutzer) [reply clang/test/Driver/XRay/xray-shared.cpp:3]: Hi, thanks for the comment. I agree with you that this check should be removed. 
The issue was pointed out to me by another person and I prepared a patch here:  #113892 (forgot to link it here before).

Comment (rupprecht) [reply clang/test/Driver/XRay/xray-shared.cpp:3]: Yes, it looks good. Thanks! Merged it in.

**Pull request merged**